## Solving QAP with Simulated Annealing Algorithm

##### AhmadReza Nopoush 610301194



In [1]:
import random
import math

at first we define function f. according to the homework f(x,y) = |sin x Cos y exp(|1 - sqrt(x^2+y^2)/pi|)

the bound of x and y is -10 < x,y < 10

In [2]:
def f(x,y) -> float:
    if abs(x) > 10 or abs(y)>10: 
        return
    return abs(math.sin(x)*math.cos(y)*math.exp(abs(1-(math.sqrt(x**2+y**2)/math.pi))))


In [3]:
class Particle_f:
    def __init__(self,lower_bound:int,upper_bound:int) -> None:
        
        #upper_bound is high limit of variable x & y
        self.upper_bound = upper_bound
        
        #lower_bound is low limit of variable x & y
        self.lower_bound = lower_bound
        
        #x is chosen randomly in bound.
        self.x = random.uniform(lower_bound,upper_bound)
        
        #y is chosen randomly in bound.
        self.y = random.uniform(lower_bound,upper_bound)
        
        #function_value keeps the value of the function. for this problem we want to maximize function_value.
        self.function_value = f(self.x,self.y)
        
        #best keeps the best solution for a particle that ever found.
        self.best = [self.x,self.y]
        
        #velocity keeps the velocity of each particle.
        self.velocity = [0,0]
        
        
    #return Velocity of a particle in a iteration.
    def Velocity(self,w,c1,c2,G_best) -> list:
        V = [0,0]
        V[0] = w*self.velocity[0] + c1*random.random()*(self.best[0]-self.x) + c2*random.random()*(G_best[0]-self.x)
        V[1] = w*self.velocity[1] + c1*random.random()*(self.best[1]-self.y) + c2*random.random()*(G_best[1]-self.y)
        return V
    
    #Updates V, x, y, function_value and best for a iteration.
    def Update(self,w,c1,c2,G_best):
        self.velocity = self.Velocity(w,c1,c2,G_best)
        self.x += self.velocity[0]
        self.y += self.velocity[1]
        
        if self.x > self.upper_bound:
            self.x = self.upper_bound
        if self.x < self.lower_bound:
            self.x = self.lower_bound
        if self.y < self.lower_bound:
            self.y = self.lower_bound
        if self.y > self.upper_bound:
            self.y = self.upper_bound
        
        if (f(self.x,self.y) > self.function_value):
            self.best = [self.x,self.y]
        self.function_value = f(self.x,self.y)

In [4]:
#main Algorithm
def PSO_find_maximum(w:float, c1:float, c2:float, population:int,iteration:int,lb:int,ub:int) -> float:
    
    #a list of particles
    Particles = [Particle_f(lb,ub) for p in range(population)]
    Particles.sort(key= lambda x: x.function_value)
    
    #G_best is best Solution that ever found.
    G_best = [Particles[-1].x,Particles[-1].y]
    
    #for each iteration do:
    for _ in range(iteration):
        
        #for each particle do:
        for p in Particles:
            p.Update(w,c1,c2,G_best)
            
        #check if a better Solution found than G_best    
        Particles.sort(key= lambda x: x.function_value)
        if f(Particles[-1].x,Particles[-1].y) > f(G_best[0],G_best[1]):
            G_best = [Particles[-1].x,Particles[-1].y]

        #print(_,Particles[-1].x,Particles[-1].y,f(Particles[-1].x,Particles[-1].y))
    return G_best,f(G_best[0],G_best[1])
    

In [5]:
PSO_find_maximum(0.5,1,0.5,100,100,-10,10)

([8.055023481807638, -9.664590029100562], 19.208502567886754)

### Results:

in 20 times of running, we got the answer ([8.055023479952993, -9.664590005991787]) with value 19.208502567886754.

notice that due to abstraction role in f, the x and y might be x = 8.05 or x = -8.05, y = -9.66 or y = 9.66



### Minimizing g

now we define function g.

the domain of g is -100 < x,y < 100

In [6]:
def g(x,y) -> float:
    if abs(x) > 100 or abs(y)>100: 
        return
    
    #I used try-exept becuase there could be division to zero.
    try:
        return (x*math.sin(math.pi*math.cos(x)*math.tan(y))*math.sin(y/x))/(1+math.cos(y/x))
    
    except:
        x+=1
        return (x*math.sin(math.pi*math.cos(x)*math.tan(y))*math.sin(y/x))/(1+math.cos(y/x))
    

In [7]:
class Particle_g:
    def __init__(self,lower_bound:int,upper_bound:int) -> None:
        
        #upper_bound is high limit of variable x & y
        self.upper_bound = upper_bound
        
        #lower_bound is low limit of variable x & y
        self.lower_bound = lower_bound
        
        #x is chosen randomly in bound.
        self.x = random.uniform(lower_bound,upper_bound)
        
        #y is chosen randomly in bound.
        self.y = random.uniform(lower_bound,upper_bound)
        
        #function_value keeps the value of the function. for this problem we want to minimize function_value.
        self.function_value = g(self.x,self.y)
        
        #best keeps the best solution for a particle that ever found.
        self.best = [self.x,self.y]
        
        #velocity keeps the velocity of each particle.
        self.velocity = [0,0]
        
    #return Velocity of a particle in a iteration.    
    def Velocity(self,w,c1,c2,G_best) -> list:
        V = [0,0]
        V[0] = w*self.velocity[0] + c1*random.random()*(self.best[0]-self.x) + c2*random.random()*(G_best[0]-self.x)
        V[1] = w*self.velocity[1] + c1*random.random()*(self.best[1]-self.y) + c2*random.random()*(G_best[1]-self.y)
        return V
    
    #Updates V, x, y, function_value and best for a iteration.
    def Update(self,w,c1,c2,G_best):
        self.velocity = self.Velocity(w,c1,c2,G_best)
        self.x += self.velocity[0]
        self.y += self.velocity[1]
        
        if self.x > self.upper_bound:
            self.x = self.upper_bound
        if self.x < self.lower_bound:
            self.x = self.lower_bound
        if self.y < self.lower_bound:
            self.y = self.lower_bound
        if self.y > self.upper_bound:
            self.y = self.upper_bound
        
        if (g(self.x,self.y) < self.function_value):
            self.best = [self.x,self.y]
        self.function_value = g(self.x,self.y)

In [8]:
#main Algoritm
def PSO_find_minimum(w:float, c1:float, c2:float, population:int,iteration:int,lb:int,ub:int) -> float:
    
    Particles = [Particle_g(lb,ub) for p in range(population)]
    Particles.sort(key= lambda x: x.function_value)
    
    #G_best is best Solution that ever found.
    G_best = [Particles[0].x,Particles[0].y]
    
    #for each iteration do:
    for _ in range(iteration):
        
        #for each particle do:
        for p in Particles:
            p.Update(w,c1,c2,G_best)
            
        #check if a better Solution found than G_best      
        Particles.sort(key= lambda x: x.function_value)
        if g(Particles[0].x,Particles[0].y) < g(G_best[0],G_best[1]):
            G_best = [Particles[0].x,Particles[0].y]

        #print(_,Particles[0].x,Particles[0].y,g(Particles[0].x,Particles[0].y))
    return G_best,g(G_best[0],G_best[1])

In [9]:
PSO_find_minimum(1,1,2,500,700,-100,100)

([-31.83098880329145, -100], -5196468637.692171)

### Results:

in 20 times of running, we got the answer [31.83098880329145, 100] with value -5196468637.692171.

notice that due to abstraction role in f, the x and y might be x = 31.83 or x = -31.83, y = -100 or y = 100

